psycopg/psycopg/connection.py

```diff
@@ -494,10 +547,72 @@ def _rollback_gen(self) -> PQGen[None]:
             return
 
         yield from self._exec_command(b"ROLLBACK")
-        cmd = self._prepared.clear()
-        if cmd:
+        self._prepared.clear()
+        for cmd in self._prepared.get_maintenance_commands():
             yield from self._exec_command(cmd)
 
+    def _pipeline_communicate_gen(self) -> PQGen[None]:
+        """Communicate with pipeline to send queries possibly fetch results,
+        which are then processing.
+        """
+        fetched = yield from pipeline_communicate(self.pgconn)
+        for results in fetched:
+            self._pipeline_process_results(results)
+
+    def _pipeline_fetch_gen(self, *, flush: bool) -> PQGen[None]:
+        """Fetch available results from the connection and process them with
+        pipeline queued items.
+
+        If 'flush' is True, a PQsendFlushRequest() is issued in order to make
+        sure results can be fetched. Otherwise, the caller may emit a
+        PQpipelineSync() call to ensure the output buffer gets flushed before
+        fetching.
+        """
+        if not self._pipeline_queue:
+            logger.debug("pipeline queue is empty, skipping fetch")
+            return
+
+        if flush:
+            logger.debug("sending flush request")
+            self.pgconn.send_flush_request()
+            yield from send(self.pgconn)
+
+        while True:
+            results = yield from fetch_many(self.pgconn)
+            if not results:
+                break
+            self._pipeline_process_results(results)
+
+    def _pipeline_process_results(self, results: List["PGresult"]) -> None:
+        """Process a results set fetched from the current pipeline.
+
+        This matchs 'results' with its respective element in the pipeline
+        queue. For commands (None value in the pipeline queue), results are
+        checked directly. For prepare statement creation requests, update the
+        cache. Otherwise, results are attached to their respective cursor.
+        """
+        queued = self._pipeline_queue.popleft()
+        logger.debug(
+            "received %s results from pipeline",
+            ", ".join(pq.ExecStatus(r.status).name for r in results),
```

Comment:
This is a relatively heavy operation which happens regardless of the log level. I think we should remove the excess of debug before merging

Reply:
I agree, they are no longer useful. Removed them, now.